src/scenic/core/visibility.py

```diff
@@ -0,0 +1,503 @@
+                h_angles_list = [np.linspace(h_range[0],h_range[1], h_ray_count) for h_ray_count in h_ray_counts]
+                angle_matrices = [np.column_stack([h_angles_list[i], np.repeat([v_angles[i]], len(h_angles_list[i]))]) for i in range(len(v_angles))]
+                angle_matrix = np.concatenate(angle_matrices, axis=0)
+            else:
+                scaled_h_ray_count = math.ceil(h_size/(viewAngles[0]) * rayCount[0])
+                h_angles = np.linspace(h_range[0],h_range[1], scaled_h_ray_count)
+                angle_matrix = np.column_stack([np.repeat(h_angles, len(v_angles)), np.tile(v_angles, len(h_angles))])
+
+            ray_vectors = np.zeros((len(angle_matrix[:,0]), 3))
+
+            ray_vectors[:,0] = -np.sin(angle_matrix[:,0])
+            ray_vectors[:,1] = np.cos(angle_matrix[:,0])
+            ray_vectors[:,2] = np.tan(angle_matrix[:,1])
+
+            ray_vectors /= np.linalg.norm(ray_vectors, axis=1)[:, np.newaxis]
+            candidate_ray_list.append(ray_vectors)
+
+        ray_vectors = np.concatenate(candidate_ray_list, axis=0)
+
+        if orientation is not None:
+            ray_vectors = orientation.getRotation().apply(ray_vectors)
+        
+        ## DEBUG ##
+        #Show all original candidate rays
+        if debug:    
+            vertices = [visibleDistance*vec + position.coordinates for vec in ray_vectors]
+            vertices = [position.coordinates] + vertices
+            lines = [trimesh.path.entities.Line([0,v]) for v in range(1,len(vertices))]
+            colors =[(0,0,255,255) for line in lines]
+
+            render_scene = trimesh.scene.Scene()
+            render_scene.add_geometry(trimesh.path.Path3D(entities=lines, vertices=vertices, process=False, colors=colors))
+            render_scene.add_geometry(target.occupiedSpace.mesh)
+            for i in list(occludingObjects):
+              render_scene.add_geometry(i.occupiedSpace.mesh)
+            render_scene.show()
+
+        # Shuffle the rays and split them into smaller batches, so we get the
+        # opportunity to return early.
+        ray_indices = np.arange(len(ray_vectors))
+        rng = np.random.default_rng(seed=42)
+        rng.shuffle(ray_indices)
+
+        batch_size = BATCH_SIZE
+
+        # Use a generator to avoid having to split a large array
+        def ray_indices_generator():
+            bottom_index = 0
+            top_index = batch_size
+
+            while bottom_index < len(ray_vectors):
+                yield ray_indices[bottom_index:top_index]
+
+                bottom_index = top_index
+                top_index += batch_size
+
+        for target_ray_indices in ray_indices_generator():
+            ray_batch = ray_vectors[target_ray_indices]
+
+            # Check if candidate rays hit target
+            raw_target_hit_info = target_region.mesh.ray.intersects_location(
+                ray_origins=[position.coordinates for ray in ray_batch],
```

Comment:
Probably faster to use `np.full` rather than a list comprehension here since the origins are all identical.